From fe11999dc659cf096f7e1fb723db34439acd23b3 Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Sat, 8 Jul 2023 16:11:00 +0800
Subject: [PATCH 08/36] [Huawei] Fix compile error when CONFIG_PSI=n
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Offering: HULK
HULK inclusion
category: bugfix
bugzilla: 188780

-------------------------------

Fix compile error when CONFIG_PSI=n

mm/vmscan.c:4412:2: error: implicit declaration of function ‘psi_memstall_type_change’; did you mean ‘psi_memstall_leave’? [-Werror=implicit-function-declaration]
  psi_memstall_type_change(PSI_GLOBAL_RECLAIM);
  ^~~~~~~~~~~~~~~~~~~~~~~~
  psi_memstall_leave
  CC      mm/mm_init.o
  AR      arch/x86/platform/efi/built-in.a
  AR      arch/x86/platform/built-in.a
mm/vmscan.c:4412:27: error: ‘PSI_GLOBAL_RECLAIM’ undeclared (first use in this function); did you mean ‘PAGEREF_RECLAIM’?
  psi_memstall_type_change(PSI_GLOBAL_RECLAIM);
                           ^~~~~~~~~~~~~~~~~~
                           PAGEREF_RECLAIM

Fixes: 994fc3d04fc61a ("PSI: add more memory fine grained stall tracking in pressure.stat")
Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 include/linux/psi.h       |  8 +++++---
 include/linux/psi_types.h | 40 +++++++++++++++++----------------------
 kernel/cgroup/cgroup.c    |  3 ++-
 3 files changed, 24 insertions(+), 27 deletions(-)

diff --git a/include/linux/psi.h b/include/linux/psi.h
index b0f6f2d0c225..afef6ba6045c 100644
--- a/include/linux/psi.h
+++ b/include/linux/psi.h
@@ -31,10 +31,8 @@ void psi_memstall_leave(unsigned long *flags);
 int psi_show(struct seq_file *s, struct psi_group *group, enum psi_res res);
 
 #ifdef CONFIG_PSI_FINE_GRAINED
-void psi_memstall_type_change(unsigned int type);
+void psi_memstall_type_change(enum psi_memstall_type type);
 int psi_stat_show(struct seq_file *s, struct psi_group *group);
-#else
-static inline void psi_memstall_type_change(unsigned int type) {}
 #endif
 
 #ifdef CONFIG_CGROUPS
@@ -73,4 +71,8 @@ static inline void cgroup_move_task(struct task_struct *p, struct css_set *to)
 
 #endif /* CONFIG_PSI */
 
+#ifndef CONFIG_PSI_FINE_GRAINED
+static inline void psi_memstall_type_change(enum psi_memstall_type type) {}
+#endif
+
 #endif /* _LINUX_PSI_H */
diff --git a/include/linux/psi_types.h b/include/linux/psi_types.h
index d61f8e25f2e2..f12425680bb8 100644
--- a/include/linux/psi_types.h
+++ b/include/linux/psi_types.h
@@ -262,23 +262,6 @@ enum psi_stat_task_count {
 	NR_PSI_STAT_TASK_COUNTS,
 };
 
-/*
- * one type should have two task stats: regular running and memstall
- * threads. The reason is the same as NR_MEMSTALL_RUNNING.
- * Because of the psi_memstall_type is start with 1, the correspondence
- * between psi_memstall_type and psi_stat_task_count should be as below:
- *
- * memstall : psi_memstall_type * 2 - 2;
- * running  : psi_memstall_type * 2 - 1;
- */
-enum psi_memstall_type {
-	PSI_MEMCG_RECLAIM = 1,
-	PSI_GLOBAL_RECLAIM,
-	PSI_COMPACT,
-	PSI_ASYNC_MEMCG_RECLAIM,
-	PSI_SWAP,
-};
-
 #define CPU_CFS_BANDWIDTH		1
 
 struct psi_group_stat_cpu {
@@ -304,12 +287,6 @@ struct psi_group_ext {
 #else
 
 struct psi_group_ext { };
-#define PSI_MEMCG_RECLAIM		0
-#define PSI_GLOBAL_RECLAIM		0
-#define PSI_COMPACT			0
-#define PSI_ASYNC_MEMCG_RECLAIM		0
-#define PSI_SWAP			0
-
 #endif /* CONFIG_PSI_FINE_GRAINED */
 
 #else /* CONFIG_PSI */
@@ -318,4 +295,21 @@ struct psi_group { };
 
 #endif /* CONFIG_PSI */
 
+/*
+ * one type should have two task stats: regular running and memstall
+ * threads. The reason is the same as NR_MEMSTALL_RUNNING.
+ * Because of the psi_memstall_type is start with 1, the correspondence
+ * between psi_memstall_type and psi_stat_task_count should be as below:
+ *
+ * memstall : psi_memstall_type * 2 - 2;
+ * running  : psi_memstall_type * 2 - 1;
+ */
+enum psi_memstall_type {
+	PSI_MEMCG_RECLAIM = 1,
+	PSI_GLOBAL_RECLAIM,
+	PSI_COMPACT,
+	PSI_ASYNC_MEMCG_RECLAIM,
+	PSI_SWAP,
+};
+
 #endif /* _LINUX_PSI_TYPES_H */
diff --git a/kernel/cgroup/cgroup.c b/kernel/cgroup/cgroup.c
index 2227bdf7c828..ed59e5c2af7f 100644
--- a/kernel/cgroup/cgroup.c
+++ b/kernel/cgroup/cgroup.c
@@ -3783,6 +3783,8 @@ static void cgroup_pressure_release(struct kernfs_open_file *of)
 	psi_trigger_destroy(ctx->psi.trigger);
 }
 
+#ifdef CONFIG_PSI_CGROUP_V1
+
 #ifdef CONFIG_PSI_FINE_GRAINED
 static int cgroup_psi_stat_show(struct seq_file *seq, void *v)
 {
@@ -3793,7 +3795,6 @@ static int cgroup_psi_stat_show(struct seq_file *seq, void *v)
 }
 #endif
 
-#ifdef CONFIG_PSI_CGROUP_V1
 struct cftype cgroup_v1_psi_files[] = {
 	{
 		.name = "io.pressure",
-- 
2.17.1

